File: server/Tingle.Dependabot/Workflow/Synchronizer.cs
﻿using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;
using System.Diagnostics.CodeAnalysis;
using Tingle.Dependabot.Events;
using Tingle.Dependabot.Models;
using Tingle.Dependabot.Models.Azure;
using Tingle.Dependabot.Models.Dependabot;
using Tingle.Dependabot.Models.Management;
using Tingle.EventBus;
using Tingle.Extensions.Primitives;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace Tingle.Dependabot.Workflow;

public interface ISynchronizer
{
    Task SynchronizeAsync(Project project, bool trigger, CancellationToken cancellationToken = default);
    Task SynchronizeAsync(Project project, Repository repository, bool trigger, CancellationToken cancellationToken = default);
    Task SynchronizeAsync(Project project, string? repositoryProviderId, bool trigger, CancellationToken cancellationToken = default);
}

internal class Synchronizer(MainDbContext dbContext,
                            IAzureDevOpsProvider adoProvider,
                            IEventPublisher publisher,
                            ILogger<Synchronizer> logger) : ISynchronizer
{
    private readonly IDeserializer yamlDeserializer = new DeserializerBuilder().WithNamingConvention(HyphenatedNamingConvention.Instance)
                                                                               .IgnoreUnmatchedProperties()
                                                                               .Build();

    public async Task SynchronizeAsync(Project project, bool trigger, CancellationToken cancellationToken = default)
    {
        // skip if the project last synchronization is less than 1 hour ago
        if ((DateTimeOffset.UtcNow - project.Synchronized) <= TimeSpan.FromHours(1))
        {
            logger.SkippingSyncProjectTooSoon(project.Id, project.Synchronized);
            return;
        }

        // update the project (it may have changed name or visibility)
        var tp = await adoProvider.GetProjectAsync(project, cancellationToken);
        var @private = tp.Visibility is not AzdoProjectVisibility.Public;
        if (!string.Equals(project.Name, tp.Name, StringComparison.Ordinal)
            || !string.Equals(project.Description, tp.Description, StringComparison.Ordinal)
            || @private != project.Private)
        {
            project.Name = tp.Name;
            project.Description = tp.Description;
            project.Private = @private;
            project.Updated = DateTimeOffset.UtcNow;
        }

        // track the synchronization pairs
        var syncPairs = new List<(SynchronizerConfigurationItem, Repository?)>();

        // get the repositories from Azure
        logger.SyncListingRepositories(project.Id);
        var adoRepos = await adoProvider.GetRepositoriesAsync(project, cancellationToken);
        logger.SyncListingRepositories(adoRepos.Count, project.Id);
        var adoReposMap = adoRepos.ToDictionary(r => r.Id.ToString(), r => r);

        // synchronize each project
        foreach (var (adoRepositoryId, adoRepo) in adoReposMap)
        {
            // skip disabled or fork repositories
            if (adoRepo.IsDisabled is true || adoRepo.IsFork)
            {
                logger.SkippingSyncRepositoryDisabledOrFork(adoRepo.Name, project.Id);
                continue;
            }

            // get the repository from the database
            var adoRepositoryName = adoRepo.Name;
            var repository = await (from r in dbContext.Repositories
                                    where r.ProjectId == project.Id
                                    where r.ProviderId == adoRepositoryId
                                    select r).SingleOrDefaultAsync(cancellationToken);

            var item = await adoProvider.GetConfigurationFileAsync(project: project,
                                                                   repositoryIdOrName: adoRepositoryId,
                                                                   cancellationToken: cancellationToken);

            // Track for further synchronization
            var sci = new SynchronizerConfigurationItem(project.Url.MakeRepositorySlug(adoRepo.Name), adoRepo, item);
            syncPairs.Add((sci, repository));
        }

        // remove repositories that are no longer tracked (i.e. the repository was removed)
        var providerIdsToKeep = syncPairs.Where(p => p.Item1.HasConfiguration).Select(p => p.Item1.Id).ToList();
        var toDeleteQuery = dbContext.Repositories.Where(r => !providerIdsToKeep.Contains(r.ProviderId!));
        var deletable = await toDeleteQuery.ToListAsync(cancellationToken);
        dbContext.RemoveRange(deletable);
        await dbContext.SaveChangesAsync(cancellationToken);
        if (deletable.Count > 0) logger.SyncDeletedRepositories(deletable.Count, project.Id);

        // synchronize each repository
        foreach (var (pi, repository) in syncPairs)
        {
            await SynchronizeAsync(project, repository, pi, trigger, cancellationToken);
        }

        // set the last synchronization time on the project
        project.Synchronized = DateTimeOffset.UtcNow;
    }

    public async Task SynchronizeAsync(Project project, Repository repository, bool trigger, CancellationToken cancellationToken = default)
    {
        // get repository
        var adoRepo = await adoProvider.GetRepositoryAsync(project: project,
                                                           repositoryIdOrName: repository.ProviderId!,
                                                           cancellationToken: cancellationToken);

        // skip disabled or fork repository
        if (adoRepo.IsDisabled is true || adoRepo.IsFork)
        {
            logger.SkippingSyncRepositoryDisabledOrFork(adoRepo.Name, project.Id);
            return;
        }

        // get the configuration file
        var item = await adoProvider.GetConfigurationFileAsync(project: project,
                                                               repositoryIdOrName: repository.ProviderId!,
                                                               cancellationToken: cancellationToken);

        // perform synchronization
        var sci = new SynchronizerConfigurationItem(project.Url.MakeRepositorySlug(adoRepo.Name), adoRepo, item);
        await SynchronizeAsync(project, repository, sci, trigger, cancellationToken);
    }

    public async Task SynchronizeAsync(Project project, string? repositoryProviderId, bool trigger, CancellationToken cancellationToken = default)
    {
        // get repository
        var adoRepo = await adoProvider.GetRepositoryAsync(project: project,
                                                           repositoryIdOrName: repositoryProviderId!,
                                                           cancellationToken: cancellationToken);

        // skip disabled or fork repository
        if (adoRepo.IsDisabled is true || adoRepo.IsFork)
        {
            logger.SkippingSyncRepositoryDisabledOrFork(adoRepo.Name, project.Id);
            return;
        }

        // get the configuration file
        var item = await adoProvider.GetConfigurationFileAsync(project: project,
                                                               repositoryIdOrName: repositoryProviderId!,
                                                               cancellationToken: cancellationToken);

        var repository = await (from r in dbContext.Repositories
                                where r.ProviderId == repositoryProviderId
                                select r).SingleOrDefaultAsync(cancellationToken);

        // perform synchronization
        var sci = new SynchronizerConfigurationItem(project.Url.MakeRepositorySlug(adoRepo.Name), adoRepo, item);
        await SynchronizeAsync(project, repository, sci, trigger, cancellationToken);
    }

    internal async Task SynchronizeAsync(Project project,
                                         Repository? repository,
                                         SynchronizerConfigurationItem providerInfo,
                                         bool trigger,
                                         CancellationToken cancellationToken = default)
    {
        // ensure not null (can be null when deleted and an event is sent)
        if (!providerInfo.HasConfiguration)
        {
            // delete repository
            if (repository is not null)
            {
                logger.SyncDeletingRepository(repositorySlug: repository.Slug, projectId: project.Id);
                dbContext.Repositories.Remove(repository);
                await dbContext.SaveChangesAsync(cancellationToken);

                // publish RepositoryDeletedEvent event
                var evt = new RepositoryDeletedEvent { ProjectId = project.Id, RepositoryId = repository.Id, };
                await publisher.PublishAsync(evt, cancellationToken: cancellationToken);
            }

            return;
        }

        // check if the file changed (different commit)
        bool commitChanged = true; // assume changes unless otherwise
        var commitId = providerInfo.CommitId;
        if (repository is not null)
        {
            commitChanged = !string.Equals(commitId, repository.LatestCommit);
        }

        // create repository
        RepositoryCreatedEvent? rce = null;
        if (repository is null)
        {
            repository = new Repository
            {
                Id = $"repo_{Ksuid.Generate()}",
                Created = DateTimeOffset.UtcNow,
                ProjectId = project.Id,
                ProviderId = providerInfo.Id,
            };
            await dbContext.Repositories.AddAsync(repository, cancellationToken);
            rce = new RepositoryCreatedEvent { ProjectId = project.Id, RepositoryId = repository.Id, };
        }

        // if the name of the repository has changed then we assume the commit changed so that we update stuff
        if (repository.Name != providerInfo.Name) commitChanged = true;

        if (commitChanged)
        {
            logger.SyncConfigFileChanged(repositorySlug: providerInfo.Slug, projectId: project.Id);

            // set/update existing values
            repository.Updated = DateTimeOffset.UtcNow;
            repository.Name = providerInfo.Name;
            repository.Slug = providerInfo.Slug;
            repository.LatestCommit = commitId;
            repository.ConfigFileContents = providerInfo.Content;

            try
            {
                var configuration = yamlDeserializer.Deserialize<DependabotConfiguration>(repository.ConfigFileContents);
                RecursiveValidator.ValidateObjectRecursive(configuration);

                // set the registries
                repository.Registries = configuration.Registries;

                // set the updates a fresh
                var updates = configuration.Updates!;
                repository.Updates = [.. updates.Select(update => new RepositoryUpdate(update)
                {
                    Files = [], // files are populated by an API call from Ruby during job execution

                    LatestJobId = null,
                    LatestJobStatus = null,
                    LatestUpdate = null,
                })];
            }
            catch (YamlDotNet.Core.YamlException ye)
            {
                logger.SyncConfigFileInvalidStructure(ye, repositorySlug: repository.Slug, projectId: project.Id);
                repository.SyncException = ye.Message;
            }
            catch (ValidationException ve)
            {
                logger.SyncConfigFileInvalidData(ve, repositorySlug: repository.Slug, projectId: project.Id);
                repository.SyncException = ve.Message;
            }

            // Update the database
            await dbContext.SaveChangesAsync(cancellationToken);

            // publish RepositoryCreatedEvent or RepositoryUpdatedEvent event
            if (rce is not null)
            {
                await publisher.PublishAsync(rce, cancellationToken: cancellationToken);
            }
            else
            {
                var evt = new RepositoryUpdatedEvent { ProjectId = project.Id, RepositoryId = repository.Id, };
                await publisher.PublishAsync(evt, cancellationToken: cancellationToken);
            }

            if (trigger)
            {
                // publish events to run update jobs for the whole repository
                var evts = repository.Updates.Select((update, index) => new RunUpdateJobEvent
                {
                    ProjectId = project.Id,
                    RepositoryId = repository.Id,
                    RepositoryUpdateId = index,
                    Trigger = UpdateJobTrigger.Synchronization,
                }).ToList();

                await publisher.PublishAsync<RunUpdateJobEvent>(evts, cancellationToken: cancellationToken);
            }
        }
    }
}

public readonly record struct SynchronizerConfigurationItem(string Id, string Name, string Slug, string? CommitId, string? Content)
{
    public SynchronizerConfigurationItem(string slug, AzdoRepository repo, AzdoRepositoryItem? item)
        : this(Id: repo.Id,
               Name: repo.Name,
               Slug: slug,
               CommitId: item?.LatestProcessedChange.CommitId,
               Content: item?.Content)
    { }

    [MemberNotNullWhen(true, nameof(CommitId))]
    [MemberNotNullWhen(true, nameof(Content))]
    public bool HasConfiguration => !string.IsNullOrEmpty(CommitId) && !string.IsNullOrEmpty(Content);
}

File: apps/web/src/workflows/sync/synchronizer.ts
import { type DependabotConfig, parseDependabotConfig } from '@paklo/core/dependabot';
import { generateId } from '@paklo/core/keygen';
import { start } from 'workflow/api';
import { logger } from '@/lib/logger';
import { type Organization, type OrganizationCredential, type Project, prisma, type Repository } from '@/lib/prisma';
import { type TriggerUpdateJobsWorkflowOptions, triggerUpdateJobs } from '@/workflows/jobs';
import { type ISyncProvider, type SynchronizerConfigurationItem, toSynchronizerProject } from './provider';

export type SyncResult = { count: number; deleted: number; updated: number };
export type SyncSingleResult = { skipped?: boolean; updated: boolean };

export class Synchronizer {
  constructor(
    readonly organization: Organization,
    readonly credential: OrganizationCredential,
    readonly provider: ISyncProvider,
    readonly project: Project,
    readonly trigger?: boolean,
  ) {}

  async syncProject(): Promise<SyncSingleResult> {
    const { project, provider } = this;

    // fetch project from provider
    const providerProj = await provider.getProject(project.providerId);
    if (!providerProj) {
      logger.warn(`Project ${project.id} not found in provider.`);
      return { updated: false };
    }

    // update project info
    logger.info(`Updating info for project ${project.id}.`);
    await prisma.project.update({
      where: { id: project.id },
      data: {
        name: providerProj.name,
        synchronizedAt: new Date(),
        synchronizationStatus: 'success',
      },
    });

    return { updated: true };
  }

  async syncRepositories(): Promise<SyncResult> {
    const { project, provider } = this;

    // track synchronization pairs
    const syncPairs: [SynchronizerConfigurationItem, Repository | null][] = [];

    // get the repositories from provider
    logger.debug(`Listing repositories in project ${project.id} ...`);
    const repos = await provider.getRepositories(project.providerId);
    if (!repos) {
      logger.debug(`No repositories found in project ${project.id}.`);
      return { count: 0, deleted: 0, updated: 0 };
    }
    logger.debug(`Found ${repos.length} repositories in ${project.id}`);
    const providerReposMap = Object.fromEntries(repos.map((r) => [r.id.toString(), r]));

    // synchronize each repository
    for (const [providerRepoId, providerRepo] of Object.entries(providerReposMap)) {
      // skip disabled or fork repositories
      if (providerRepo.disabled || providerRepo.fork) {
        logger.debug(`Skipping sync for ${providerRepo.name} in ${project.id} because it is disabled or is a fork`);
        continue;
      }

      // get the repository from the database
      const repository = await prisma.repository.findUnique({
        where: {
          projectId_providerId: { projectId: project.id, providerId: providerRepoId },
        },
      });

      // get the configuration file
      const sci = await provider.getConfigurationFile(toSynchronizerProject(project), providerRepo);

      // track for further synchronization
      syncPairs.push([sci, repository]);

      // add delay of 30ms to avoid rate limiting, every 10 repositories
      if (syncPairs.length % 10 === 0) {
        await new Promise((resolve) => setTimeout(resolve, 30));
      }
    }

    // remove repositories that are no longer tracked (i.e. the repository was removed)
    const providerIdsToKeep = syncPairs.filter(([item]) => item.hasConfiguration).map(([item]) => item.id);
    const { count: deleted } = await prisma.repository.deleteMany({
      where: {
        projectId: project.id,
        providerId: { notIn: providerIdsToKeep },
      },
    });
    logger.info(`Deleted ${deleted} repositories in project ${project.id} that no longer have configuration files.`);

    // synchronize each repository
    let updated = 0;
    for (const [sci, repository] of syncPairs) {
      const { updated: repoUpdated } = await this.synchronizeInner({
        project,
        repository,
        providerInfo: sci,
      });
      if (repoUpdated) updated++;
    }

    return { count: repos.length, deleted, updated };
  }

  async syncRepo(params: { repository: Repository }): Promise<SyncSingleResult> {
    const { project, provider } = this;
    const { repository } = params;

    // get repository from provider
    const providerRepo = await provider.getRepository(project.providerId, repository.providerId);
    if (!providerRepo) {
      logger.warn(`Repository ${repository.providerId} not found in project ${project.id}.`);
      return { updated: false };
    }

    // skip disabled or fork repository
    if (providerRepo.disabled || providerRepo.fork) {
      logger.info(`Skipping sync for ${providerRepo.name} in ${project.id} because it is disabled or is a fork`);
      return { updated: false };
    }

    // get the configuration file
    const sci = await provider.getConfigurationFile(toSynchronizerProject(project), providerRepo);

    // perform synchronization
    return await this.synchronizeInner({
      project,
      repository,
      providerInfo: sci,
    });
  }

  async syncRepoByProvider(params: { repositoryProviderId: string }): Promise<SyncSingleResult> {
    const { project, provider } = this;
    const { repositoryProviderId } = params;

    // get repository from provider
    const providerRepo = await provider.getRepository(project.providerId, repositoryProviderId);
    if (!providerRepo) {
      logger.warn(`Repository ${repositoryProviderId} not found in project ${project.id}.`);
      return { updated: false };
    }

    // skip disabled or fork repository
    if (providerRepo.disabled || providerRepo.fork) {
      logger.info(`Skipping sync for ${providerRepo.name} in ${project.id} because it is disabled or is a fork`);
      return { updated: false };
    }

    // get the configuration file
    const sci = await provider.getConfigurationFile(toSynchronizerProject(project), providerRepo);

    // get the repository from the database
    const repository = await prisma.repository.findUnique({
      where: {
        projectId_providerId: { projectId: project.id, providerId: providerRepo.id },
      },
    });

    // perform synchronization
    return await this.synchronizeInner({
      project,
      repository,
      providerInfo: sci,
    });
  }

  private async synchronizeInner({
    project,
    repository,
    providerInfo,
  }: {
    project: Project;
    repository: Repository | null;
    providerInfo: SynchronizerConfigurationItem;
  }): Promise<SyncSingleResult> {
    // ensure not null (can be null when deleted and an event is sent)
    if (!providerInfo.hasConfiguration) {
      // delete repository
      if (repository) {
        logger.info(`Deleting '${repository.slug}' in ${project.id} as it no longer has a configuration file.`);
        await prisma.repository.delete({ where: { id: repository.id } });
      }

      return { updated: false };
    }

    // check if the file changed (different commit)
    let commitChanged: boolean = true; // assume changes unless otherwise
    const commitId = providerInfo.commitId;
    if (repository) {
      commitChanged = commitId !== repository.latestCommit;
    }

    // create repository
    if (!repository) {
      repository = await prisma.repository.create({
        data: {
          id: generateId(),
          projectId: project.id,
          providerId: providerInfo.id,
          name: providerInfo.name,
          slug: providerInfo.slug,
          url: providerInfo.url,
          permalink: providerInfo.permalink,
          latestCommit: commitId,
          configFileContents: providerInfo.content,
          synchronizationStatus: 'pending',
        },
      });
    }

    // if the name/slug/url of the repository has changed then we assume the commit changed so that we update stuff
    if (
      repository.name !== providerInfo.name ||
      repository.slug !== providerInfo.slug ||
      repository.url !== providerInfo.url
    ) {
      commitChanged = true;
    }

    if (!commitChanged) return { updated: false };

    // at this point we know the commit or info changed
    logger.info(`Changes detected for repository '${providerInfo.slug}' in project ${project.id} ...`);

    // update repository info
    let _config: DependabotConfig;
    let syncError: string | null = null;
    try {
      _config = await parseDependabotConfig({
        configContents: providerInfo.content!,
        configPath: providerInfo.path!,
        variableFinder: () => undefined,
      });
    } catch (error) {
      syncError = (error as Error).message;
    }

    repository = await prisma.repository.update({
      where: { id: repository.id },
      data: {
        name: providerInfo.name,
        slug: providerInfo.slug,
        url: providerInfo.url,
        latestCommit: commitId,
        configFileContents: providerInfo.content,
        synchronizationStatus: syncError ? 'failed' : 'success',
        synchronizationError: syncError,
        synchronizedAt: new Date(),
      },
    });

    // TODO: save updates from parsed config into own table/collection

    if (this.trigger) {
      // trigger update jobs for the whole repository
      await start(triggerUpdateJobs, [
        {
          organizationId: this.organization.id,
          projectId: project.id,
          repositoryId: repository.id,
          repositoryUpdateId: undefined, // run all
          trigger: 'synchronization',
        } satisfies TriggerUpdateJobsWorkflowOptions,
      ]);
    }

    return { updated: true };
  }
}
